File: ts/test/entity/calendar/CalendarDirect.test.ts

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'


import { ChurchCalendarSDK } from '../../..'

import {
  envOverride,
  liveDelay,
  maybeSkipControl,
  skipIfMissingIds,
} from '../../utility'


describe('CalendarDirect', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when CHURCHCALENDAR_TEST_LIVE=TRUE.
  afterEach(liveDelay('CHURCHCALENDAR_TEST_LIVE'))

  test('direct-exists', async () => {
    const sdk = new ChurchCalendarSDK({
      system: { fetch: async () => ({}) }
    })
    assert('function' === typeof sdk.direct)
    assert('function' === typeof sdk.prepare)
  })


  test('direct-list-calendar', async (t: any) => {
    const setup = directSetup([{ id: 'direct01' }, { id: 'direct02' }])
    if (maybeSkipControl(t, 'direct', 'direct-list-calendar', setup.live)) return
    if (skipIfMissingIds(t, setup, ["calendar01","day01","month01","year01"])) return
    const { client, calls } = setup

    const params: any = {}
    const query: any = {}
    if (setup.live) {
      params.calendar = setup.idmap['calendar01']
      params.day = setup.idmap['day01']
      params.month = setup.idmap['month01']
      params.year = setup.idmap['year01']
    } else {
      params.calendar = 'direct01'
      params.day = 'direct02'
      params.month = 'direct03'
      params.year = 'direct04'
    }

    const result: any = await client.direct({
      path: 'api/v0/en/calendars/{calendar}/{year}/{month}/{day}',
      method: 'GET',
      params,
      query,
    })

    if (setup.live) {
      // Live mode is lenient: synthetic IDs frequently 4xx and the list-
      // response shape varies wildly across public APIs. Skip rather than
      // fail when the call doesn't return a usable list.
      if (!result.ok || result.status < 200 || result.status >= 300) {
        return
      }
      const listArr = unwrapListData(result.data)
      if (!Array.isArray(listArr)) {
        return
      }
    } else {
      assert(result.ok === true)
      assert(result.status === 200)
      assert(null != result.data)
      const listArr = unwrapListData(result.data)
      assert(Array.isArray(listArr))
      assert(listArr!.length === 2)
      assert(calls.length === 1)
      assert(calls[0].init.method === 'GET')
      assert(calls[0].url.includes('direct01'))
      assert(calls[0].url.includes('direct02'))
      assert(calls[0].url.includes('direct03'))
      assert(calls[0].url.includes('direct04'))
    }
  })

})



function directSetup(mockres?: any) {
  const calls: any[] = []

  const env = envOverride({
    'CHURCHCALENDAR_TEST_CALENDAR_ENTID': {},
    'CHURCHCALENDAR_TEST_LIVE': 'FALSE',
  })

  const live = 'TRUE' === env.CHURCHCALENDAR_TEST_LIVE

  if (live) {
    const client = new ChurchCalendarSDK({
    })

    let idmap: any = env['CHURCHCALENDAR_TEST_CALENDAR_ENTID']
    if ('string' === typeof idmap && idmap.startsWith('{')) {
      idmap = JSON.parse(idmap)
    }

    return { client, calls, live, idmap }
  }

  const mockFetch = async (url: string, init: any) => {
    calls.push({ url, init })
    return {
      status: 200,
      statusText: 'OK',
      headers: {},
      json: async () => (null != mockres ? mockres : { id: 'direct01' }),
    }
  }

  const client = new ChurchCalendarSDK({
    base: 'http://localhost:8080',
    system: { fetch: mockFetch },
  })

  return { client, calls, live, idmap: {} as any }
}

// direct() returns the raw response body. List endpoints often wrap the
// array in an envelope (e.g. { data: [...] }, { entities: [...] },
// { pagination, data: [...] }). The test transforms the raw body to
// extract the first array — either the body itself or the first array
// property of an envelope object.
function unwrapListData(data: any): any[] | null {
  if (Array.isArray(data)) return data
  if (data && 'object' === typeof data) {
    for (const v of Object.values(data)) {
      if (Array.isArray(v)) return v as any[]
    }
  }
  return null
}
  

File: rb/test/calendar_direct_test.rb
# Calendar direct test

require "minitest/autorun"
require "json"
require_relative "../ChurchCalendar_sdk"
require_relative "runner"

class CalendarDirectTest < Minitest::Test
  def test_direct_list_calendar
    setup = calendar_direct_setup([
      { "id" => "direct01" },
      { "id" => "direct02" },
    ])
    _should_skip, _reason = Runner.is_control_skipped("direct", "direct-list-calendar", setup[:live] ? "live" : "unit")
    if _should_skip
      skip(_reason || "skipped via sdk-test-control.json")
      return
    end
    if setup[:live]
      ["calendar01", "day01", "month01", "year01"].each do |_live_key|
        if setup[:idmap][_live_key].nil?
          skip "live test needs #{_live_key} via *_ENTID env var (synthetic IDs only)"
          return
        end
      end
    end
    client = setup[:client]

    params = {}
    if setup[:live]
      params["calendar"] = setup[:idmap]["calendar01"]
    else
      params["calendar"] = "direct01"
    end
    if setup[:live]
      params["day"] = setup[:idmap]["day01"]
    else
      params["day"] = "direct01"
    end
    if setup[:live]
      params["month"] = setup[:idmap]["month01"]
    else
      params["month"] = "direct01"
    end
    if setup[:live]
      params["year"] = setup[:idmap]["year01"]
    else
      params["year"] = "direct01"
    end

    result = client.direct({
      "path" => "api/v0/en/calendars/{calendar}/{year}/{month}/{day}",
      "method" => "GET",
      "params" => params,
    })
    if setup[:live]
      # Live mode is lenient: synthetic IDs frequently 4xx and the list-
      # response shape varies wildly across public APIs. Skip rather than
      # fail when the call doesn't return a usable list.
      if !result["err"].nil?
        skip("list call failed (likely synthetic IDs against live API): #{result["err"]}")
        return
      end
      unless result["ok"]
        skip("list call not ok (likely synthetic IDs against live API)")
        return
      end
      status = Helpers.to_int(result["status"])
      if status < 200 || status >= 300
        skip("expected 2xx status, got #{status}")
        return
      end
    else
      assert_nil result["err"]
      assert result["ok"]
      assert_equal 200, Helpers.to_int(result["status"])
      assert result["data"].is_a?(Array)
      assert_equal 2, result["data"].length
      assert_equal 1, setup[:calls].length
    end
  end

end


def calendar_direct_setup(mockres)
  Runner.load_env_local

  calls = []

  env = Runner.env_override({
    "CHURCHCALENDAR_TEST_CALENDAR_ENTID" => {},
    "CHURCHCALENDAR_TEST_LIVE" => "FALSE",
  })

  live = env["CHURCHCALENDAR_TEST_LIVE"] == "TRUE"

  if live
    merged_opts = {
    }
    client = ChurchCalendarSDK.new(merged_opts)
    return {
      client: client,
      calls: calls,
      live: true,
      idmap: {},
    }
  end

  mock_fetch = ->(url, init) {
    calls.push({ "url" => url, "init" => init })
    return {
      "status" => 200,
      "statusText" => "OK",
      "headers" => {},
      "json" => ->() {
        if !mockres.nil?
          return mockres
        end
        return { "id" => "direct01" }
      },
      "body" => "mock",
    }, nil
  }

  client = ChurchCalendarSDK.new({
    "base" => "http://localhost:8080",
    "system" => {
      "fetch" => mock_fetch,
    },
  })

  {
    client: client,
    calls: calls,
    live: false,
    idmap: {},
  }
end
